pkg/puller/downloader.go

```diff
@@ -0,0 +1,172 @@
+package puller
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"hash/fnv"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+var (
+	d *Downloader
+)
+
+type Downloader struct {
+	NumRetries int
+	ModelDir   string
+}
+
+type Protocol string
+
+const (
+	S3    Protocol = "s3://"
+	GCS    Protocol = "gs://"
+	//PVC   Protocol = "pvc://"
+	//File  Protocol = "file://"
+	//HTTPS Protocol = "https://"
+)
+
+// TODO: Add more supported protocols
+var SupportedProtocols = []Protocol{S3, GCS}
+
+func init() {
+	d = NewDownloader()
+}
+
+func NewDownloader() *Downloader {
+	d := new(Downloader)
+	return d
+}
+
+func SetDownloadSettings(numRetries int, modelDir string) {
+	d.NumRetries = numRetries
+	d.ModelDir = modelDir
+}
+
+func DownloadModel(event EventWrapper) error {
+	modelSpec := event.ModelSpec
+	if modelSpec != nil {
+		modelUri := modelSpec.StorageURI
+		hashString := hash(modelUri)
+		log.Println("Processing:", modelUri, "=", hashString)
+		successFile := filepath.Join(d.ModelDir, fmt.Sprintf("SUCCESS.%d", hashString))
+		ok := fileExists(successFile)
+		if !ok {
+			downloadErr := download(modelUri)
+			if downloadErr != nil {
+				return fmt.Errorf("download error: %v", downloadErr)
+			} else {
+				file, createErr := os.Create(successFile)
+				if createErr != nil {
+					return fmt.Errorf("create file error: %v", createErr)
```

Comment:
consider an annotation pattern that will yield cleaner error messages:

"while creating file: %v"